Use coordinate transforms to run in position

GiudGiud opened the pull request:
refs #19121

I thought I needed this to debug general field transfers but turns out not really. As a result this is lower priority

Thoughts?

Todo:
- [x] Test this using all the coordinate transforms tests, using custom comparison files that ignore the mesh coordinates and only compare the fields
- [x] Check if output_in_position could be extended to coordinate transforms

EDIT: output in position is done simply by passing a point. To extend it we would need to pass a coordinate transform. It s not that difficult, just not the focus for now. 

EDIT: only did two tests from the coordinate transforms. The bulk of these tests are with every kind of transfer. I think we are covered already with the ones we have

GiudGiud committed:
Add capability to transform the mesh by the coordinate transform

GiudGiud committed:
Move translation to app coord transform to be able to set it from the [Mesh] block

GiudGiud committed:
Add parameter to run in position for the multiapp
refs #19121

lindsayad reviewed (commented):
This seems pretty reasonable to me. It would be nice to see the coordinate transformations be actually useful haha

GiudGiud committed:
Avoid performing null transformations on mesh
Add missing parenthesis
Fix error messages

refs #19121

GiudGiud committed:
Add testing to running MultiApps in position
refs #19121

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1321576/) on 439a417 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22866/site)

This comment will be updated on new commits.

GiudGiud committed:
Place errors at the right locations for unsupported features
Set the translation as early as possible for the coordinate transform
refs #19121

GiudGiud commented:
@lindsayad this one ready as well, though not high priority

GiudGiud commented:
nevermind broke regular transfers.

GiudGiud committed:
Re-add translations to multiapp coord transform for transfers

Fix issues in minimal data unpacking
- set euler angles
- always compute the _rs matrix, even if no rotation
- unpack the right member for translation

refs #19121

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1323204/) on 92e8ef9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22866/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22866/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1584ac2ec2d98e7f4756228788341656aa44cfd`

GiudGiud commented:
ok this should pass but I dodged a difficulty (below) by duplicating an attribute and not sure that will be ok with you
- handling the constness of the origin app transform when setting the translation after the multiAppCoordTransform has been created

the good news is that the last commit has a bug fix for a corner case: scaling with no rotation and multiple subapps

GiudGiud commented on framework/src/utils/MooseAppCoordTransform.C:
```diff
@@ -204,6 +215,7 @@ MooseAppCoordTransform::validParams()
       "coord_type", coord_types, "Type of the coordinate system per block param");
   params.addParam<MooseEnum>(
       "rz_coord_axis", rz_coord_axis, "The rotation axis (X | Y) for axisymetric coordinates");
+  params.addParam<Point>("translation", "Translation vector. Override by the app position");
```
this is up for deletion. It would be a way to define the position in the subapp input file. 

It s not compatible with the positions argument of the MultiApp
It also only works with run_in_position right now

lindsayad replied:
I would turn the question around and ask if you have interest in having this? On the one hand, it would nice to have some symmetry with the other types of transform. On the other if there's no feature advantage to this over `positions`, then I don't care to have multiple ways to do the same thing generally

GiudGiud replied:
not really. i am doing this for symmetry

GiudGiud replied:
let s delete it. My desire for symmetry isnt worth handling an extra parameter

GiudGiud commented:
Ready again. Since we no longer keep a refernce to the original transform, we could drop the constness instead of the const_cast, lemme know

GiudGiud commented:
Is this better
MultiAppCoordTransform::MultiAppCoordTransform(const MooseAppCoordTransform & our_app_transform)
or this
MultiAppCoordTransform::MultiAppCoordTransform(const MooseAppCoordTransform our_app_transform)
??

the second one is clear we are not saving a reference so for the dev who doesnt look at the whole class it's clearer

the first one only creates one copy of the parameter (_our_app_transform is not a reference anymore), I think the second creates a copy when dealing with the parameter then another to save in the class member

GiudGiud commented:
Still on the fence with these last two commits. Seems like a shame to loose the const-ness. 

Maybe the code is better off two commits ago with a simple check that _translation is always equal to _moose_app_coord._translation.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1336299/) on bbba4c0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>11b387</th>
      <th colspan=4>#22866 bbba4c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.70%</td>
      <td>84.71%</td>
      <td>+0.01%</td>
      <td>89.23%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82845</td>
      <td>82893</td>
      <td>+48</td>
      <td>58</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14965</td>
      <td>14963</td>
      <td>-2</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22866/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22866/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22866/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.23% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
How about I move translation to be an argument to transformMesh

then we dont need it as a class member and we can keep all the const-ness we had

GiudGiud committed:
Remove translation attribution from MooseAppCoord, position will remain a data known by the MultiApp, and the MultiAppTransform, rather than breaking the paradigm and making it a mooseapp data

GiudGiud commented:
ok now it s ready

lindsayad reviewed (changes requested):
this came out nice and clean IMO

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Should be addressed

GiudGiud commented:
Thanks for the review!
